src/cudamatrix/cu-kernels.cu

```diff
@@ -251,6 +251,48 @@ static void _select_rows(const int* out_row_ptr, int* out_col_idx,
   }
 }
 
+// mat += alpha * smat
+//
+// We use warpSize threads per row to access only the nonzero elements.
+// Every CU1DBLOCK/warpSize rows share one thread block.
+// 1D grid to cover all rows of smat.
+template<typename Real>
+__global__
+static void _add_smat(Real* mat, MatrixDim mat_dim, Real alpha,
+                      const int* smat_row_ptr, const int* smat_col_idx,
+                      const Real* smat_val) {
+  const int i = blockIdx.x * blockDim.y + threadIdx.y; // row idx
+  if (i < mat_dim.rows) {
```

Comment:
It's not clear to me that this code would be very efficient in the case where there's exactly one nonzero element per row, which is quite common in practice.  It seems like it would launch an excessive number of thread blocks.
 I'm concerned that for the regular neural net training, where we use this function, we could experience a slowdown.  
One option here is to call the function csr2coo() with a temporarily allocated piece of memory, and then make a kernel that uses the coo format, which will be much more efficient.  Then we can rely on the NVidia people having implemented csr2coo() in a smart way.
Let me know what you think.